core/shared/src/test/scala/fs2/StreamSpec.scala

```diff
@@ -914,6 +914,91 @@ class StreamSpec extends Fs2Spec {
       }
     }
 
+    "evalFilter" - {
+      "with effectful const(true)" in forAll { (s: Stream[Pure, Int]) =>
+        withClue("takes all elements") {
+          s.evalFilter(_ => IO.pure(true)).compile.toList.asserting(_ shouldBe s.toList)
+        }
+      }
+
+      "with effectful const(false)" in forAll { (s: Stream[Pure, Int]) =>
+        withClue("takes no elements") {
+          s.evalFilter(_ => IO.pure(false)).compile.toList.asserting(_ shouldBe empty)
+        }
+      }
+
+      "with function that filters out odd elements" in {
+        Stream
+          .range(1, 10)
+          .evalFilter(e => IO(e % 2 == 0))
+          .compile
+          .toList
+          .asserting(_ shouldBe List(2, 4, 6, 8))
+      }
+    }
+
+    "evalFilterAsync" - {
+      "with effectful const(true)" in forAll { (s: Stream[Pure, Int]) =>
+        withClue("takes all elements") {
+          s.covary[IO]
+            .evalFilterAsync(5)(_ => IO.pure(true))
+            .compile
+            .toList
+            .asserting(_ shouldBe s.toList)
+        }
+      }
+
+      "with effectful const(false)" in forAll { (s: Stream[Pure, Int]) =>
+        withClue("takes no elements") {
+          s.covary[IO]
+            .evalFilterAsync(5)(_ => IO.pure(false))
+            .compile
+            .toList
+            .asserting(_ shouldBe empty)
+        }
+      }
+
+      "with function that filters out odd elements" in {
+        Stream
+          .range(1, 10)
+          .evalFilterAsync[IO](5)(e => IO(e % 2 == 0))
+          .compile
+          .toList
+          .asserting(_ shouldBe List(2, 4, 6, 8))
+      }
+
+      "filters up to N items in parallel" in {
```

Comment:
This one was tough... I wanted to avoid using `sleep` to ensure some parallelism is achieved, but I couldn't figure out a better way.

Possibly I could avoid using the semaphore (by failing from `sig.modify` if the value is too large) or the signal (by getting the amount of available permits, but that'd require looping the check in `Stream.repeatEval`). But I guess since this is test code we don't need to care about perf *that* much.

Unless there's a reliable way to ensure that the filtering action waits for a few elements to start processing - that'd speed up the test and probably ensure its determinism as well...